File: lib/features/insights/domain/insight_engine.dart
import 'dart:math' as math;

import '../../cycle/domain/cycle_models.dart';
import '../../predictions/domain/prediction_history.dart';

class CycleSummary {
  const CycleSummary({
    required this.averageLength,
    required this.shortestLength,
    required this.longestLength,
    required this.standardDeviation,
  });

  final double averageLength;
  final int shortestLength;
  final int longestLength;
  final double standardDeviation;
}

class InsightSnapshot {
  const InsightSnapshot({
    required this.cycleSummary,
    required this.averagePeriodDurationDays,
    required this.flowCounts,
    required this.observationCounts,
    required this.observationalMessages,
    this.predictionEvaluation,
  });

  final CycleSummary? cycleSummary;
  final double? averagePeriodDurationDays;
  final Map<FlowLevel, int> flowCounts;
  final Map<ObservationKind, int> observationCounts;
  final List<String> observationalMessages;
  final PredictionEvaluation? predictionEvaluation;
}

class InsightEngine {
  CycleSummary? summarizeCycleLengths(List<int> lengths) {
    if (lengths.isEmpty) return null;
    final average = lengths.reduce((a, b) => a + b) / lengths.length;
    final variance = lengths
            .map((v) => math.pow(v - average, 2).toDouble())
            .reduce((a, b) => a + b) /
        lengths.length;
    return CycleSummary(
      averageLength: average,
      shortestLength: lengths.reduce(math.min),
      longestLength: lengths.reduce(math.max),
      standardDeviation: math.sqrt(variance),
    );
  }

  InsightSnapshot summarize({
    required List<PeriodEpisode> periods,
    required List<HealthObservation> observations,
    PredictionEvaluation? predictionEvaluation,
  }) {
    final ordered = [...periods]..sort((a, b) => a.start.compareTo(b.start));
    final cycleLengths = <int>[];
    for (var i = 1; i < ordered.length; i++) {
      final days = _date(ordered[i].start).difference(_date(ordered[i - 1].start)).inDays;
      if (days >= 15 && days <= 90) cycleLengths.add(days);
    }

    final completedDurations = ordered.map((item) => item.durationDays).whereType<int>().where((value) => value > 0).toList();
    final averageDuration = completedDurations.isEmpty
        ? null
        : completedDurations.reduce((a, b) => a + b) / completedDurations.length;

    final flowCounts = <FlowLevel, int>{};
    final observationCounts = <ObservationKind, int>{};
    for (final observation in observations) {
      observationCounts.update(observation.kind, (value) => value + 1, ifAbsent: () => 1);
      final flow = observation.flowLevel;
      if (flow != null) flowCounts.update(flow, (value) => value + 1, ifAbsent: () => 1);
    }

    final messages = <String>[];
    final timingKinds = <ObservationKind, String>{
      ObservationKind.cramps: 'Cramps',
      ObservationKind.headache: 'Headaches',
      ObservationKind.migraine: 'Migraines',
      ObservationKind.mood: 'Mood observations',
      ObservationKind.sleep: 'Sleep observations',
      ObservationKind.energy: 'Energy observations',
      ObservationKind.stress: 'Stress observations',
    };
    for (final entry in timingKinds.entries) {
      final days = _cycleDaysFor(entry.key, ordered, observations);
      if (days.length >= 2) {
        messages.add(symptomTimingInsight(symptomLabel: entry.value, cycleDays: days));
      }
    }
    if (flowCounts.isNotEmpty) {
      final top = flowCounts.entries.reduce((a, b) => a.value >= b.value ? a : b);
      messages.add('${_title(top.key.name)} flow was your most frequently recorded flow level (${top.value} logs). This is a summary of your entries, not a medical interpretation.');
    }

    return InsightSnapshot(
      cycleSummary: summarizeCycleLengths(cycleLengths),
      averagePeriodDurationDays: averageDuration,
      flowCounts: Map.unmodifiable(flowCounts),
      observationCounts: Map.unmodifiable(observationCounts),
      observationalMessages: List.unmodifiable(messages),
      predictionEvaluation: predictionEvaluation,
    );
  }

  String symptomTimingInsight({
    required String symptomLabel,
    required List<int> cycleDays,
  }) {
    if (cycleDays.isEmpty) {
      return '$symptomLabel has not been recorded often enough for a timing insight.';
    }
    final counts = <int, int>{};
    for (final day in cycleDays) {
      counts.update(day, (v) => v + 1, ifAbsent: () => 1);
    }
    final top = counts.entries.reduce((a, b) => a.value >= b.value ? a : b);
    return '$symptomLabel were recorded most often around cycle day ${top.key}. This is an observation, not a medical cause or diagnosis.';
  }

  List<int> _cycleDaysFor(
    ObservationKind kind,
    List<PeriodEpisode> periods,
    List<HealthObservation> observations,
  ) {
    if (periods.isEmpty) return const [];
    final result = <int>[];
    for (final observation in observations.where((item) => item.kind == kind)) {
      PeriodEpisode? latest;
      for (final period in periods) {
        if (!_date(period.start).isAfter(_date(observation.occurredAt))) latest = period;
      }
      if (latest == null) continue;
      final day = _date(observation.occurredAt).difference(_date(latest.start)).inDays + 1;
      if (day >= 1 && day <= 90) result.add(day);
    }
    return result;
  }

  DateTime _date(DateTime value) => DateTime(value.year, value.month, value.day);

  String _title(String value) => value.isEmpty ? value : '${value[0].toUpperCase()}${value.substring(1)}';
}

File: web/src/features/insights/insight-engine.ts
import type { PredictionEvaluation } from '../../domain/prediction/prediction-history';
import type {
  FlowLevel,
  HealthObservation,
  ObservationKind,
  PeriodEpisode,
  RecordSource,
} from '../../domain/cycle/types';

const DAY_MS = 86_400_000;

export const INSIGHT_CAPABILITY_IDS = [
  'INS-001', 'INS-002', 'INS-003', 'INS-004',
  'INS-005', 'INS-006', 'INS-007', 'INS-008',
  'INS-009', 'INS-010', 'INS-011', 'INS-012',
  'INS-013', 'INS-014', 'INS-015', 'INS-016',
] as const;

export type CycleSummary = {
  averageLength: number;
  shortestLength: number;
  longestLength: number;
  standardDeviation: number;
};

export type InsightProvenance = {
  sources: RecordSource[];
  dateRange: { from: string; to: string } | null;
};

export type TimingInsight = {
  kind: ObservationKind;
  cycleDay: number;
};

export type PmsPatternInsight = {
  kind: ObservationKind;
  count: number;
};

export type FlowPatternInsight = {
  flow: FlowLevel;
  count: number;
};

export type InsightSnapshot = {
  cycleSummary: CycleSummary | null;
  averagePeriodDurationDays: number | null;
  flowCounts: Partial<Record<FlowLevel, number>>;
  observationCounts: Partial<Record<ObservationKind, number>>;
  timingInsights: TimingInsight[];
  predictionEvaluation?: PredictionEvaluation;
  cycleLengths: number[];
  pmsPatterns: PmsPatternInsight[];
  topFlowPattern: FlowPatternInsight | null;
  provenance: InsightProvenance;
};

function utcDateMs(value: string): number {
  const [year, month, day] = value.slice(0, 10).split('-').map(Number);
  return Date.UTC(year, month - 1, day);
}

function periodDurationDays(period: PeriodEpisode): number | null {
  if (period.end == null) return null;
  return Math.floor((utcDateMs(period.end) - utcDateMs(period.start)) / DAY_MS) + 1;
}

function summarizeCycleLengths(lengths: number[]): CycleSummary | null {
  if (lengths.length === 0) return null;
  const averageLength = lengths.reduce((sum, value) => sum + value, 0) / lengths.length;
  const variance = lengths.reduce(
    (sum, value) => sum + ((value - averageLength) ** 2),
    0,
  ) / lengths.length;

  return {
    averageLength,
    shortestLength: Math.min(...lengths),
    longestLength: Math.max(...lengths),
    standardDeviation: Math.sqrt(variance),
  };
}

function cycleDaysFor(
  kind: ObservationKind,
  periods: PeriodEpisode[],
  observations: HealthObservation[],
): number[] {
  if (periods.length === 0) return [];
  const result: number[] = [];

  for (const observation of observations.filter((item) => item.kind === kind)) {
    let latest: PeriodEpisode | undefined;
    const observationDay = utcDateMs(observation.occurredAt);
    for (const period of periods) {
      if (utcDateMs(period.start) <= observationDay) latest = period;
    }
    if (!latest) continue;
    const day = Math.floor((observationDay - utcDateMs(latest.start)) / DAY_MS) + 1;
    if (day >= 1 && day <= 90) result.push(day);
  }

  return result;
}

function mostFrequentCycleDay(cycleDays: number[]): number | null {
  if (cycleDays.length === 0) return null;

  const counts = new Map<number, number>();
  for (const day of cycleDays) counts.set(day, (counts.get(day) ?? 0) + 1);
  let topDay = cycleDays[0];
  let topCount = counts.get(topDay) ?? 0;
  for (const [day, count] of counts) {
    if (count > topCount) {
      topDay = day;
      topCount = count;
    }
  }
  return topDay;
}

function buildProvenance(
  periods: PeriodEpisode[],
  observations: HealthObservation[],
): InsightProvenance {
  const sources: RecordSource[] = [];
  const rememberSource = (source: RecordSource) => {
    if (!sources.includes(source)) sources.push(source);
  };
  periods.forEach((period) => rememberSource(period.source));
  observations.forEach((observation) => rememberSource(observation.source));

  const timestamps = [
    ...periods.flatMap((period) => period.end == null ? [period.start] : [period.start, period.end]),
    ...observations.map((observation) => observation.occurredAt),
  ];
  if (timestamps.length === 0) return { sources, dateRange: null };
  const ordered = [...timestamps].sort((left, right) => Date.parse(left) - Date.parse(right));
  return {
    sources,
    dateRange: { from: ordered[0], to: ordered[ordered.length - 1] },
  };
}

export function summarizeInsights({
  periods,
  observations,
  predictionEvaluation,
}: {
  periods: PeriodEpisode[];
  observations: HealthObservation[];
  predictionEvaluation?: PredictionEvaluation;
}): InsightSnapshot {
  const ordered = [...periods].sort((left, right) => Date.parse(left.start) - Date.parse(right.start));
  const cycleLengths: number[] = [];
  for (let index = 1; index < ordered.length; index += 1) {
    const days = Math.floor((utcDateMs(ordered[index].start) - utcDateMs(ordered[index - 1].start)) / DAY_MS);
    if (days >= 15 && days <= 90) cycleLengths.push(days);
  }

  const completedDurations = ordered
    .map(periodDurationDays)
    .filter((value): value is number => value !== null && value > 0);
  const averagePeriodDurationDays = completedDurations.length === 0
    ? null
    : completedDurations.reduce((sum, value) => sum + value, 0) / completedDurations.length;

  const flowCounts: Partial<Record<FlowLevel, number>> = {};
  const observationCounts: Partial<Record<ObservationKind, number>> = {};
  for (const observation of observations) {
    observationCounts[observation.kind] = (observationCounts[observation.kind] ?? 0) + 1;
    if (observation.flowLevel) {
      flowCounts[observation.flowLevel] = (flowCounts[observation.flowLevel] ?? 0) + 1;
    }
  }

  const timingInsights: TimingInsight[] = [];
  const timingKinds: ObservationKind[] = [
    'cramps',
    'headache',
    'migraine',
    'mood',
    'sleep',
    'energy',
    'stress',
  ];
  for (const kind of timingKinds) {
    const days = cycleDaysFor(kind, ordered, observations);
    if (days.length < 2) continue;
    const cycleDay = mostFrequentCycleDay(days);
    if (cycleDay !== null) timingInsights.push({ kind, cycleDay });
  }

  const pmsKinds = new Set<ObservationKind>([
    'cramps',
    'bloating',
    'breastTenderness',
    'headache',
    'migraine',
    'fatigue',
    'mood',
    'irritability',
  ]);
  const pmsCounts = new Map<ObservationKind, number>();
  for (const observation of observations) {
    if (!pmsKinds.has(observation.kind)) continue;
    const observationDay = utcDateMs(observation.occurredAt);
    const nextPeriod = ordered.find((period) => utcDateMs(period.start) > observationDay);
    if (!nextPeriod) continue;
    const daysBefore = Math.round((utcDateMs(nextPeriod.start) - observationDay) / DAY_MS);
    if (daysBefore >= 1 && daysBefore <= 7) {
      pmsCounts.set(observation.kind, (pmsCounts.get(observation.kind) ?? 0) + 1);
    }
  }
  const pmsPatterns = [...pmsCounts.entries()].map(([kind, count]) => ({ kind, count }));

  const flowEntries = Object.entries(flowCounts) as Array<[FlowLevel, number]>;
  let topFlowPattern: FlowPatternInsight | null = null;
  if (flowEntries.length > 0) {
    let top = flowEntries[0];
    for (const entry of flowEntries.slice(1)) {
      if (entry[1] > top[1]) top = entry;
    }
    topFlowPattern = { flow: top[0], count: top[1] };
  }

  return {
    cycleSummary: summarizeCycleLengths(cycleLengths),
    averagePeriodDurationDays,
    flowCounts,
    observationCounts,
    timingInsights,
    ...(predictionEvaluation ? { predictionEvaluation } : {}),
    cycleLengths,
    pmsPatterns,
    topFlowPattern,
    provenance: buildProvenance(ordered, observations),
  };
}
